2023-07-03-ruby-regex/us/lib/state_machine.rb

```diff
@@ -1,20 +1,83 @@
 # frozen_string_literal: true
 
 class StateMachine
+  attr_reader :graph, :opts
+
   def initialize(transitions, opts)
-    @graph = transitions.each_line.each_with_object({}) do |line, g|
+    graph_init = Hash.new { |h, k| h[k] = transition_hash }
+    graph_init[opts[:start]] = transition_hash
+    graph_init[opts[:finish]] = transition_hash
+    @graph = transitions.each_line.each_with_object(graph_init) do |line, g|
       state1, edge, state2 = line.split(/ /)
-      g[state1.to_i] ||= {}
-      g[state1.to_i][edge] ||= []
       g[state1.to_i][edge].push(state2.to_i)
     end
     @opts = opts
   end
 
   def run(input)
-    input.chars.inject(epsilon_steps([@opts[:start]])) do |acc, c|
+    input.chars.inject(epsilon_steps([opts[:start]])) do |acc, c|
       epsilon_steps(step_fw(acc, c))
-    end.include?(@opts[:finish])
+    end.include?(opts[:finish])
+  end
+
+  def concat(other)
+    # finish of self to start of other
+    max_state = graph.keys.max
+    other.opts
+    other.graph.each do |state, edges|
+      new_state =
+        if state == other.opts[:start]
+          opts[:finish]
+        else
+          state + max_state
+        end
+      edges.each do |edge, next_states|
+        next_states.each do |next_state|
+          new_next_state =
+            if next_state == other.opts[:start]
+              opts[:finish]
+            else
+              next_state + max_state
+            end
```

Comment:
@josh-hamacher-cobalt I stil want to refactor these.
In EU time zone I did a lambda, but I don't like it that much.